src/dbnode/encoding/proto/encoder.go

```diff
@@ -0,0 +1,910 @@
+	// fashion if we write our own decoder or expose the one in the underlying library.
+	if err := enc.unmarshaled.Unmarshal(protoBytes); err != nil {
+		return fmt.Errorf(
+			"%s error unmarshaling annotation into proto message: %v", encErrPrefix, err)
+	}
+
+	if len(enc.unmarshaled.GetUnknownFields()) > 0 {
+		// TODO(rartoul): Make this behavior configurable / this may change once we implement
+		// mid-stream schema changes to make schema upgrades easier for clients.
+		// https://github.com/m3db/m3/issues/1471
+		return errEncoderMessageHasUnknownFields
+	}
+
+	// From this point onwards all errors are "hard errors" meaning that they should render
+	// the encoder unsable since we may have encoded partial data.
+
+	if enc.numEncoded == 0 {
+		enc.encodeHeader()
+	}
+
+	if timeUnit != enc.timestampEncoder.TimeUnit {
+		// We handle encoding time unit changes ourselves because by default the WriteTime()
+		// API will use a marker encoding scheme which relies on looking ahead into the stream
+		// for bit combinations that could not possibly exist in the M3TSZ encoding scheme. We don't
+		// want to rely on this behavior because its possible that we could encode a legit set of bits
+		// that matches the "impossible" M3TSZ markers exactly.
+
+		// First bit means either there is no more data OR the time unit has changed.
+		enc.stream.WriteBit(opCodeNoMoreDataOrTimeUnitChange)
+		// Next bit means there is more data, but the time unit has changed.
+		enc.stream.WriteBit(opCodeTimeUnitChange)
+
+		enc.timestampEncoder.WriteTimeUnit(enc.stream, timeUnit)
+	} else {
+		// Control bit that indicates the stream has more data.
+		enc.stream.WriteBit(opCodeMoreData)
+	}
+
+	err := enc.timestampEncoder.WriteTime(enc.stream, dp.Timestamp, nil, timeUnit)
+	if err != nil {
+		enc.hardErr = err
+		return fmt.Errorf(
+			"%s error encoding timestamp: %v", encErrPrefix, err)
+	}
+
+	if err := enc.encodeProto(enc.unmarshaled); err != nil {
+		enc.hardErr = err
+		return fmt.Errorf(
+			"%s error encoding proto portion of message: %v", encErrPrefix, err)
+	}
+
+	enc.numEncoded++
+	enc.lastEncodedDP = dp
+	return nil
+}
+
+// Stream returns a copy of the underlying data stream.
+func (enc *Encoder) Stream() xio.SegmentReader {
+	seg := enc.segment(true)
+	if seg.Len() == 0 {
+		return nil
+	}
+
+	if readerPool := enc.opts.SegmentReaderPool(); readerPool != nil {
+		reader := readerPool.Get()
+		reader.Reset(seg)
+		return reader
+	}
+	return xio.NewSegmentReader(seg)
+}
+
+func (enc *Encoder) segment(copy bool) ts.Segment {
+	length := enc.stream.Len()
+	if enc.stream.Len() == 0 {
+		return ts.Segment{}
+	}
+
+	var head checked.Bytes
+	buffer, _ := enc.stream.Rawbytes()
+	if !copy {
+		// Take ref from the ostream.
+		head = enc.stream.Discard()
+	} else {
+		// Copy into new buffer.
+		head = enc.newBuffer(length)
+		head.IncRef()
+		head.AppendAll(buffer)
+		head.DecRef()
+	}
+
+	return ts.NewSegment(head, nil, ts.FinalizeHead)
```

Comment:
Hm, you never have to calculate the tail here?  Are there unused bits at the end of the stream?  How do those correctly get not looked at?  

Or do you rewrite bits in the tail?  (i.e. go backwards when you have more data and delete the trailing marker of no more data?)

Reply:
Discussed offline, don't need the tail here because of the existing control bits. Added an extra item to this issue to track the matter of maybe adding a tail: https://github.com/m3db/m3/issues/1471